### DISCUSSION TITLE
How to read a file on the server side?

### DISCUSSION THREAD

**Discussion by ghost** [nuxt/nuxt#17591] [Questions]
Hello, I don't understand how to read a file on the server side (/server/api).
I'm trying to do IP geolocation to convert it into a country with MaxMind.

And I have to do this:
`maxmind.open("@/server/GeoLite2-Country.mmdb");`

I tried all the paths, ~/assets/ etc but it never finds the file since nuxt is built in a .nuxt folder

Any idea? Is it not possible?

**Comment by UnderKoen** [comment]
Have you tried using a absolute path? Like `C:\Users\[user]\projects\example\server\geo.mmdb`

If this works you can use `path.resolve` to find the absolute file.

like:
```ts
import path from 'path';

const mmdb = path.resolve("server/GeoLite2-Country.mmdb");
maxmind.open(mmdb);
```

**Comment by ghost** [comment]
@UnderKoen  Yes, when I write the absolute path it works of course. But I'd like it to work with the relative path so that it works with different OS like windows, linux and docker. If I understand correctly, when you build nuxt it creates an .output folder and puts everything in it, except my file

**Reply by UnderKoen** [reply]
@antoineanorra I updated my orginal answer to also work with those cases

**Reply by ghost** [reply]
Yes it works in npm run dev but not in production mode. This is the principle of the .nuxt / .output folder. As my file is not cloned in these folders, it may not be a best practice, to be confirmed, maybe by @danielroe who will give us a real solution

**Comment by viii47** [comment]
Hey @antoineanorra,

I added a `data` folder to my current project (unit testing fake-server-data) which are mostly JSON-Files. I suppose this should work for a `*.mmdb` file as well.

![image](https://user-images.githubusercontent.com/95022556/168770266-5db94b1b-0e8f-47a2-9d21-29b5e00e95bc.png)

```ts

/**
 * testing fake-data
 * server/api/search/test.ts
 */
import searchresult from '~/server/data/unitTest.search.result.json';

export default defineEventHandler(async (event) => { 
    const body = await useBody(event);
    const queryValue = body.query.toString().toLowerCase();
    const catValue = body.cat;
    console.log(searchresult);
    return searchresult.data
});

```

**Reply by ghost** [reply]
Hey @viii47, can you show me with which line of code you read this file? Have you tried it in production?

**Reply by viii47** [reply]
I'll have a look if it still works as-is in rc.3x

**Reply by viii47** [reply]
@antoineanorra I updated my answer with how the api looks like. I also confirmed that it works in a production environment. Maybe there's a better way how to include third party data into the server part, though this seems to work as intended.

**Reply by ghost** [reply]
Yes with import keyword your file is cloned so that’s way it’s works. But there is a limitation in file extension

**Comment by UnderKoen** [comment]
You can use nitro assets by placing your file in "server/assets". 

Then you can acces it by the storage of nitro.
Like:
```ts
// @ts-ignore
import {useStorage} from '#imports';
import type { Storage } from 'unstorage'

const storage = useStorage() as Storage;
```

```ts
//Getting the content
await storage.getItem("assets:server:[filename]");
```

**Reply by ghost** [reply]
useStorage doesn't exist in #imports, maybe it comes from another lib. And the problem is that with your method, I have to read the content, which bothers me since I'm just trying to get the path to open the file and then the maxmind lib reads the file partially (faster).

**Reply by ghost** [reply]
If there is no direct solution, maybe I should copy this file myself in the build folder (.nuxt/.output) to be able to use it as an absolute path like you suggested with the resolve method

**Reply by UnderKoen** [reply]
It does exist, because you can run it without errors, it think nitro imports aren't included in type correction for the editor.

Futher if you run from the root directory the resolve works, you would need to run the production server with: "node .ouput/server/index.mjs"

An other possibilty would be to add a custom rollup/vite building plugin, but I dont know how to achieve this currently

**Comment by cardin** [comment]
This is a little late, but:

```ts
// https://nitro.unjs.io/guide/storage#configuration
// nuxt.config.ts
export default defineNuxtConfig({
  nitro: {
    storage: {
      projectDir: {
        driver: "fsLite",
        base: ".",
        readOnly: true,
      },
    },
  },
});
```

```ts
// https://unstorage.unjs.io/drivers/fs
// server/utils/db/Database.ts
const createTableText = await useStorage("projectDir")
    .getItem("server/utils/db/DbCreate.sql")
    .then((x) => x?.toString());
```
